src/socket/encrypter.rs

```diff
@@ -0,0 +1,250 @@
+    send_incoming: Sender<HavenMsg>,
+    _task: Arc<Task<anyhow::Result<()>>>,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub enum HavenMsg {
+    ClientHs(Handshake),
+    ServerHs(Handshake),
+    Regular { nonce: u64, inner: Bytes },
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct Handshake {
+    id_pk: IdentityPublic,
+    eph_pk: OnionPublic,
+    sig: Bytes,
+}
+
+impl Encrypter {
+    pub fn client_new(
+        my_isk: IdentitySecret,
+        remote_ep: Endpoint,
+        rendezvous_fp: Option<Fingerprint>,
+        n2r_skt: N2rSocket,
+        send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
+        ctx: DaemonContext,
+    ) -> Self {
+        let (send_out, recv_out) = smol::channel::bounded(1000);
+        let (send_in, recv_in) = smol::channel::bounded(1000);
+        let task = smolscale::spawn(enc_task(
+            my_isk,
+            n2r_skt,
+            remote_ep,
+            rendezvous_fp,
+            recv_in,
+            recv_out,
+            send_incoming_decrypted,
+            None,
+            ctx,
+        ));
+        Self {
+            send_outgoing: send_out,
+            send_incoming: send_in,
+            _task: Arc::new(task),
+        }
+    }
+
+    pub fn server_new(
+        my_isk: IdentitySecret,
+        remote_ep: Endpoint,
+        rendezvous_fp: Option<Fingerprint>,
+        n2r_skt: N2rSocket,
+        send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
+        client_hs: Handshake,
+        ctx: DaemonContext,
+    ) -> Self {
+        let (send_out, recv_out) = smol::channel::bounded(1000);
+        let (send_in, recv_in) = smol::channel::bounded(1000);
+        let task = smolscale::spawn(enc_task(
+            my_isk,
+            n2r_skt,
+            remote_ep,
+            rendezvous_fp,
+            recv_in,
+            recv_out,
+            send_incoming_decrypted,
+            Some(client_hs),
+            ctx,
+        ));
+        Self {
+            send_outgoing: send_out,
+            send_incoming: send_in,
+            _task: Arc::new(task),
+        }
+    }
+
+    pub async fn send_outgoing(&self, msg: Bytes) -> Result<(), SocketSendError> {
+        self.send_outgoing
+            .send(msg)
+            .await
+            .map_err(|_| SocketSendError::HavenSendError)
+    }
+
+    pub async fn send_incoming(&self, msg: HavenMsg) -> Result<(), SocketRecvError> {
+        self.send_incoming
+            .send(msg)
+            .await
+            .map_err(|_| SocketRecvError::HavenRecvError)
+    }
+}
+
+async fn enc_task(
+    my_isk: IdentitySecret,
+    n2r_skt: N2rSocket,
+    remote_ep: Endpoint,
+    rendezvous_fp: Option<Fingerprint>,
+    recv_incoming: Receiver<HavenMsg>,
+    recv_outgoing: Receiver<Bytes>,
+    send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
+    client_hs: Option<Handshake>,
+    ctx: DaemonContext,
+) -> anyhow::Result<()> {
+    // get rendezvous ep
+    let rendezvous_ep = match rendezvous_fp {
```

Comment:
The whole point of moving this into the task is to do the lookup *every single packet*. We eventually might even support rendezvous load-balancing, where a single haven has multiple rendezvous, between which we randomly pick when we look up!

This is best done by simply writing a local `send_to_rendezvous` helper function and calling it down below.